File: ios/PhoneRadar/Services/BLEManager.swift
//
//  BLEManager.swift
//  PhoneRadar
//
//  Created on 2024-01-09.
//

import Foundation
import CoreBluetooth
import Combine

// MARK: - BLE Service UUIDs
struct BLEServiceUUIDs {
    // Standard Services
    static let battery = CBUUID(string: "0000180F-0000-1000-8000-00805F9B34FB")
    static let deviceInfo = CBUUID(string: "0000180A-0000-1000-8000-00805F9B34FB")

    // Custom Services
    static let radarData = CBUUID(string: "12345678-1234-5678-1234-56789ABCDEF0")
    static let thermalData = CBUUID(string: "12345678-1234-5678-1234-56789ABCDEF1")
    static let sensorData = CBUUID(string: "12345678-1234-5678-1234-56789ABCDEF2")
}

struct BLECharacteristicUUIDs {
    // Battery
    static let batteryLevel = CBUUID(string: "00002A19-0000-1000-8000-00805F9B34FB")

    // Device Info
    static let manufacturerName = CBUUID(string: "00002A29-0000-1000-8000-00805F9B34FB")
    static let modelNumber = CBUUID(string: "00002A24-0000-1000-8000-00805F9B34FB")
    static let firmwareRevision = CBUUID(string: "00002A26-0000-1000-8000-00805F9B34FB")

    // Radar
    static let radarDistance = CBUUID(string: "12345678-1234-5678-1234-56789ABCDEF3")
    static let radarImage = CBUUID(string: "12345678-1234-5678-1234-56789ABCDEF4")

    // Thermal
    static let thermalImage = CBUUID(string: "12345678-1234-5678-1234-56789ABCDEF5")
    static let ambientTemp = CBUUID(string: "12345678-1234-5678-1234-56789ABCDEF6")

    // Sensor
    static let signalStrength = CBUUID(string: "12345678-1234-5678-1234-56789ABCDEF7")
}

// MARK: - Models
struct RadarDeviceInfo {
    var name: String
    var manufacturer: String?
    var model: String?
    var firmware: String?
}

struct SensorData {
    var batteryLevel: Int = 0
    var temperature: Float = 0.0
    var distance: Float = 0.0
    var signalStrength: Int = 0
    var timestamp: Date = Date()
}

struct RadarImageData {
    var width: Int
    var height: Int
    var pixels: [UInt8]
    var timestamp: Date
}

// MARK: - BLE Manager
class BLEManager: NSObject, ObservableObject {
    // Published properties
    @Published var isScanning = false
    @Published var isConnected = false
    @Published var discoveredDevices: [CBPeripheral] = []
    @Published var deviceInfo = RadarDeviceInfo(name: "")
    @Published var sensorData = SensorData()
    @Published var radarImage: RadarImageData?
    @Published var thermalImage: RadarImageData?
    @Published var errorMessage: String?

    // Private properties
    private var centralManager: CBCentralManager!
    private var connectedPeripheral: CBPeripheral?
    private var characteristics: [CBUUID: CBCharacteristic] = [:]

    override init() {
        super.init()
        centralManager = CBCentralManager(delegate: self, queue: .main)
    }

    // MARK: - Public Methods
    func startScanning() {
        guard centralManager.state == .poweredOn else {
            errorMessage = "Bluetooth açık değil"
            return
        }

        discoveredDevices.removeAll()
        isScanning = true

        centralManager.scanForPeripherals(
            withServices: [BLEServiceUUIDs.battery],
            options: [CBCentralManagerScanOptionAllowDuplicatesKey: false]
        )

        // 10 saniye sonra taramayı durdur
        DispatchQueue.main.asyncAfter(deadline: .now() + 10) { [weak self] in
            self?.stopScanning()
        }
    }

    func stopScanning() {
        centralManager.stopScan()
        isScanning = false
    }

    func connect(to peripheral: CBPeripheral) {
        stopScanning()
        connectedPeripheral = peripheral
        peripheral.delegate = self
        centralManager.connect(peripheral, options: nil)
    }

    func disconnect() {
        if let peripheral = connectedPeripheral {
            centralManager.cancelPeripheralConnection(peripheral)
        }
    }

    // MARK: - Private Methods
    private func discoverServices() {
        guard let peripheral = connectedPeripheral else { return }

        peripheral.discoverServices([
            BLEServiceUUIDs.battery,
            BLEServiceUUIDs.deviceInfo,
            BLEServiceUUIDs.radarData,
            BLEServiceUUIDs.thermalData,
            BLEServiceUUIDs.sensorData
        ])
    }

    private func processData(from characteristic: CBCharacteristic) {
        guard let data = characteristic.value else { return }

        switch characteristic.uuid {
        case BLECharacteristicUUIDs.batteryLevel:
            if let batteryLevel = data.first {
                sensorData.batteryLevel = Int(batteryLevel)
            }

        case BLECharacteristicUUIDs.manufacturerName:
            deviceInfo.manufacturer = String(data: data, encoding: .utf8)

        case BLECharacteristicUUIDs.modelNumber:
            deviceInfo.model = String(data: data, encoding: .utf8)

        case BLECharacteristicUUIDs.firmwareRevision:
            deviceInfo.firmware = String(data: data, encoding: .utf8)

        case BLECharacteristicUUIDs.radarDistance:
            if data.count >= 4 {
                let distance = data.withUnsafeBytes { $0.load(as: Float.self) }
                sensorData.distance = distance
            }

        case BLECharacteristicUUIDs.ambientTemp:
            if data.count >= 4 {
                let temp = data.withUnsafeBytes { $0.load(as: Float.self) }
                sensorData.temperature = temp
            }

        case BLECharacteristicUUIDs.radarImage:
            parseImageData(data: data, isRadar: true)

        case BLECharacteristicUUIDs.thermalImage:
            parseImageData(data: data, isRadar: false)

        default:
            break
        }
    }

    private func parseImageData(data: Data, isRadar: Bool) {
        guard data.count >= 4 else { return }

        let width = Int(data[0]) | (Int(data[1]) << 8)
        let height = Int(data[2]) | (Int(data[3]) << 8)

        let pixels = Array(data[4...])

        let imageData = RadarImageData(
            width: width,
            height: height,
            pixels: pixels,
            timestamp: Date()
        )

        if isRadar {
            radarImage = imageData
        } else {
            thermalImage = imageData
        }
    }
}

// MARK: - CBCentralManagerDelegate
extension BLEManager: CBCentralManagerDelegate {
    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        switch central.state {
        case .poweredOn:
            errorMessage = nil
        case .poweredOff:
            errorMessage = "Bluetooth kapalı"
        case .unauthorized:
            errorMessage = "Bluetooth izni verilmemiş"
        case .unsupported:
            errorMessage = "Bu cihaz Bluetooth desteklemiyor"
        default:
            errorMessage = "Bluetooth durumu bilinmiyor"
        }
    }

    func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String : Any], rssi RSSI: NSNumber) {
        // PhoneRadar veya FLIR başlangıcıyla cihazları filtrele
        if let name = peripheral.name,
           (name.hasPrefix("PhoneRadar") || name.hasPrefix("FLIR") || name.hasPrefix("Radar")) {
            if !discoveredDevices.contains(where: { $0.identifier == peripheral.identifier }) {
                discoveredDevices.append(peripheral)
            }
        }
    }

    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        isConnected = true
        deviceInfo.name = peripheral.name ?? "Bilinmeyen Cihaz"
        discoverServices()
    }

    func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        isConnected = false
        connectedPeripheral = nil
        characteristics.removeAll()

        if let error = error {
            errorMessage = "Bağlantı kesildi: \(error.localizedDescription)"
        }
    }

    func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
        errorMessage = "Bağlanılamadı: \(error?.localizedDescription ?? "Bilinmeyen hata")"
    }
}

// MARK: - CBPeripheralDelegate
extension BLEManager: CBPeripheralDelegate {
    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        if let error = error {
            errorMessage = "Servisler bulunamadı: \(error.localizedDescription)"
            return
        }

        guard let services = peripheral.services else { return }

        for service in services {
            peripheral.discoverCharacteristics(nil, for: service)
        }
    }

    func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        if let error = error {
            print("Karakteristikler bulunamadı: \(error.localizedDescription)")
            return
        }

        guard let characteristics = service.characteristics else { return }

        for characteristic in characteristics {
            self.characteristics[characteristic.uuid] = characteristic

            // Read initial values
            if characteristic.properties.contains(.read) {
                peripheral.readValue(for: characteristic)
            }

            // Enable notifications
            if characteristic.properties.contains(.notify) {
                peripheral.setNotifyValue(true, for: characteristic)
            }
        }
    }

    func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
        if let error = error {
            print("Değer okunamadı: \(error.localizedDescription)")
            return
        }

        processData(from: characteristic)
    }
}

File: android/app/src/main/java/com/phoneradar/app/services/BLEManager.kt
package com.phoneradar.app.services

import android.annotation.SuppressLint
import android.bluetooth.*
import android.bluetooth.le.BluetoothLeScanner
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import android.content.Context
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.*

// MARK: - BLE Service UUIDs
object BLEServiceUUIDs {
    // Standard Services
    val BATTERY = UUID.fromString("0000180F-0000-1000-8000-00805F9B34FB")
    val DEVICE_INFO = UUID.fromString("0000180A-0000-1000-8000-00805F9B34FB")

    // Custom Services
    val RADAR_DATA = UUID.fromString("12345678-1234-5678-1234-56789ABCDEF0")
    val THERMAL_DATA = UUID.fromString("12345678-1234-5678-1234-56789ABCDEF1")
    val SENSOR_DATA = UUID.fromString("12345678-1234-5678-1234-56789ABCDEF2")
}

object BLECharacteristicUUIDs {
    // Battery
    val BATTERY_LEVEL = UUID.fromString("00002A19-0000-1000-8000-00805F9B34FB")

    // Device Info
    val MANUFACTURER_NAME = UUID.fromString("00002A29-0000-1000-8000-00805F9B34FB")
    val MODEL_NUMBER = UUID.fromString("00002A24-0000-1000-8000-00805F9B34FB")
    val FIRMWARE_REVISION = UUID.fromString("00002A26-0000-1000-8000-00805F9B34FB")

    // Radar
    val RADAR_DISTANCE = UUID.fromString("12345678-1234-5678-1234-56789ABCDEF3")
    val RADAR_IMAGE = UUID.fromString("12345678-1234-5678-1234-56789ABCDEF4")

    // Thermal
    val THERMAL_IMAGE = UUID.fromString("12345678-1234-5678-1234-56789ABCDEF5")
    val AMBIENT_TEMP = UUID.fromString("12345678-1234-5678-1234-56789ABCDEF6")

    // Sensor
    val SIGNAL_STRENGTH = UUID.fromString("12345678-1234-5678-1234-56789ABCDEF7")
}

// MARK: - Data Models
data class RadarDeviceInfo(
    val name: String = "",
    val manufacturer: String? = null,
    val model: String? = null,
    val firmware: String? = null
)

data class SensorData(
    val batteryLevel: Int = 0,
    val temperature: Float = 0f,
    val distance: Float = 0f,
    val signalStrength: Int = 0,
    val timestamp: Long = System.currentTimeMillis()
)

data class RadarImageData(
    val width: Int,
    val height: Int,
    val pixels: ByteArray,
    val timestamp: Long = System.currentTimeMillis()
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as RadarImageData

        if (width != other.width) return false
        if (height != other.height) return false
        if (!pixels.contentEquals(other.pixels)) return false

        return true
    }

    override fun hashCode(): Int {
        var result = width
        result = 31 * result + height
        result = 31 * result + pixels.contentHashCode()
        return result
    }
}

// MARK: - BLE Manager
@SuppressLint("MissingPermission")
class BLEManager(private val context: Context) {

    private val bluetoothAdapter: BluetoothAdapter? = BluetoothAdapter.getDefaultAdapter()
    private val bluetoothLeScanner: BluetoothLeScanner? = bluetoothAdapter?.bluetoothLeScanner

    private var bluetoothGatt: BluetoothGatt? = null
    private val characteristics = mutableMapOf<UUID, BluetoothGattCharacteristic>()

    // State flows
    private val _isScanning = MutableStateFlow(false)
    val isScanning: StateFlow<Boolean> = _isScanning.asStateFlow()

    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected.asStateFlow()

    private val _discoveredDevices = MutableStateFlow<List<BluetoothDevice>>(emptyList())
    val discoveredDevices: StateFlow<List<BluetoothDevice>> = _discoveredDevices.asStateFlow()

    private val _deviceInfo = MutableStateFlow(RadarDeviceInfo())
    val deviceInfo: StateFlow<RadarDeviceInfo> = _deviceInfo.asStateFlow()

    private val _sensorData = MutableStateFlow(SensorData())
    val sensorData: StateFlow<SensorData> = _sensorData.asStateFlow()

    private val _radarImage = MutableStateFlow<RadarImageData?>(null)
    val radarImage: StateFlow<RadarImageData?> = _radarImage.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    // MARK: - Public Methods
    fun startScanning() {
        if (!isBluetoothEnabled()) {
            _errorMessage.value = "Bluetooth açık değil"
            return
        }

        _discoveredDevices.value = emptyList()
        _isScanning.value = true

        bluetoothLeScanner?.startScan(scanCallback)
    }

    fun stopScanning() {
        bluetoothLeScanner?.stopScan(scanCallback)
        _isScanning.value = false
    }

    fun connect(device: BluetoothDevice) {
        stopScanning()
        bluetoothGatt = device.connectGatt(context, false, gattCallback)
    }

    fun disconnect() {
        bluetoothGatt?.disconnect()
        bluetoothGatt?.close()
        bluetoothGatt = null
        _isConnected.value = false
        characteristics.clear()
    }

    fun isBluetoothEnabled(): Boolean {
        return bluetoothAdapter?.isEnabled == true
    }

    // MARK: - Scan Callback
    private val scanCallback = object : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            val device = result.device
            val deviceName = device.name

            // Filter for PhoneRadar, FLIR, or Radar devices
            if (deviceName != null &&
                (deviceName.startsWith("PhoneRadar") ||
                        deviceName.startsWith("FLIR") ||
                        deviceName.startsWith("Radar"))
            ) {
                val currentDevices = _discoveredDevices.value.toMutableList()
                if (!currentDevices.any { it.address == device.address }) {
                    currentDevices.add(device)
                    _discoveredDevices.value = currentDevices
                }
            }
        }

        override fun onScanFailed(errorCode: Int) {
            _errorMessage.value = "Tarama başarısız: $errorCode"
            _isScanning.value = false
        }
    }

    // MARK: - GATT Callback
    private val gattCallback = object : BluetoothGattCallback() {
        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            when (newState) {
                BluetoothProfile.STATE_CONNECTED -> {
                    _isConnected.value = true
                    _deviceInfo.value = RadarDeviceInfo(name = gatt.device.name ?: "Bilinmeyen Cihaz")
                    gatt.discoverServices()
                }
                BluetoothProfile.STATE_DISCONNECTED -> {
                    _isConnected.value = false
                    _errorMessage.value = "Bağlantı kesildi"
                }
            }
        }

        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
            if (status == BluetoothGatt.GATT_SUCCESS) {
                gatt.services.forEach { service ->
                    service.characteristics.forEach { characteristic ->
                        characteristics[characteristic.uuid] = characteristic

                        // Read initial values
                        if (characteristic.properties and BluetoothGattCharacteristic.PROPERTY_READ != 0) {
                            gatt.readCharacteristic(characteristic)
                        }

                        // Enable notifications
                        if (characteristic.properties and BluetoothGattCharacteristic.PROPERTY_NOTIFY != 0) {
                            gatt.setCharacteristicNotification(characteristic, true)

                            // Write descriptor to enable notifications
                            characteristic.descriptors.forEach { descriptor ->
                                if (descriptor.uuid == UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")) {
                                    descriptor.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                                    gatt.writeDescriptor(descriptor)
                                }
                            }
                        }
                    }
                }
            }
        }

        override fun onCharacteristicRead(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
            status: Int
        ) {
            if (status == BluetoothGatt.GATT_SUCCESS) {
                processCharacteristic(characteristic)
            }
        }

        override fun onCharacteristicChanged(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic
        ) {
            processCharacteristic(characteristic)
        }
    }

    // MARK: - Data Processing
    private fun processCharacteristic(characteristic: BluetoothGattCharacteristic) {
        val data = characteristic.value ?: return

        when (characteristic.uuid) {
            BLECharacteristicUUIDs.BATTERY_LEVEL -> {
                if (data.isNotEmpty()) {
                    _sensorData.value = _sensorData.value.copy(
                        batteryLevel = data[0].toInt() and 0xFF
                    )
                }
            }

            BLECharacteristicUUIDs.MANUFACTURER_NAME -> {
                _deviceInfo.value = _deviceInfo.value.copy(
                    manufacturer = String(data, Charsets.UTF_8)
                )
            }

            BLECharacteristicUUIDs.MODEL_NUMBER -> {
                _deviceInfo.value = _deviceInfo.value.copy(
                    model = String(data, Charsets.UTF_8)
                )
            }

            BLECharacteristicUUIDs.FIRMWARE_REVISION -> {
                _deviceInfo.value = _deviceInfo.value.copy(
                    firmware = String(data, Charsets.UTF_8)
                )
            }

            BLECharacteristicUUIDs.RADAR_DISTANCE -> {
                if (data.size >= 4) {
                    val distance = ByteBuffer.wrap(data)
                        .order(ByteOrder.LITTLE_ENDIAN)
                        .float
                    _sensorData.value = _sensorData.value.copy(distance = distance)
                }
            }

            BLECharacteristicUUIDs.AMBIENT_TEMP -> {
                if (data.size >= 4) {
                    val temp = ByteBuffer.wrap(data)
                        .order(ByteOrder.LITTLE_ENDIAN)
                        .float
                    _sensorData.value = _sensorData.value.copy(temperature = temp)
                }
            }

            BLECharacteristicUUIDs.RADAR_IMAGE -> {
                parseImageData(data, isRadar = true)
            }

            BLECharacteristicUUIDs.THERMAL_IMAGE -> {
                parseImageData(data, isRadar = false)
            }
        }
    }

    private fun parseImageData(data: ByteArray, isRadar: Boolean) {
        if (data.size < 4) return

        val buffer = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
        val width = buffer.getShort(0).toInt() and 0xFFFF
        val height = buffer.getShort(2).toInt() and 0xFFFF

        val pixels = data.copyOfRange(4, data.size)

        val imageData = RadarImageData(
            width = width,
            height = height,
            pixels = pixels
        )

        if (isRadar) {
            _radarImage.value = imageData
        }
    }
}
